src/Features/Core/Portable/Completion/CompletionList.cs

```diff
@@ -1,43 +1,136 @@
+            this.SuggestionModeItem = suggestionModeItem;
         }
+
+        /// <summary>
+        /// Creates a new <see cref="CompletionList"/> instance.
+        /// </summary>
+        /// <param name="defaultSpan">The span of the syntax element at the caret position when the <see cref="CompletionList"/> was created.</param>
+        /// <param name="items">The completion items to present to the user.</param>
+        /// <param name="rules">The rules used to control behavior of the completion list shown to the user during typing.</param>
+        /// <param name="suggestionModeItem">An optional <see cref="CompletionItem"/> that appears selected in the list presented to the user during suggestion mode.</param>
+        /// <returns></returns>
+        public static CompletionList Create(
+            TextSpan defaultSpan,
+            ImmutableArray<CompletionItem> items,
+            CompletionRules rules = null,
+            CompletionItem suggestionModeItem = null)
+        {
+            return new CompletionList(defaultSpan, FixItemSpans(items, defaultSpan), rules, suggestionModeItem);
+        }
+
+        private static ImmutableArray<CompletionItem> FixItemSpans(ImmutableArray<CompletionItem> items, TextSpan defaultSpan)
+        {
+            if (defaultSpan != default(TextSpan) && items.Any(i => i.Span == default(TextSpan)))
+            {
+                items = items.Select(i => i.Span == default(TextSpan) ? i.WithSpan(defaultSpan) : i).ToImmutableArray();
+            }
+
+            return items;
+        }
+
+        private CompletionList With(
+            Optional<TextSpan> span = default(Optional<TextSpan>),
+            Optional<ImmutableArray<CompletionItem>> items = default(Optional<ImmutableArray<CompletionItem>>),
+            Optional<CompletionRules> rules = default(Optional<CompletionRules>),
+            Optional<CompletionItem> suggestionModeItem = default(Optional<CompletionItem>))
+        {
+            var newSpan = span.HasValue ? span.Value : this.DefaultSpan;
+            var newItems = items.HasValue ? items.Value : this.Items;
+            var newRules = rules.HasValue ? rules.Value : this.Rules;
+            var newSuggestionModeItem = suggestionModeItem.HasValue ? suggestionModeItem.Value : this.SuggestionModeItem;
+
+            if (newSpan == this.DefaultSpan 
+                && newItems == this.Items 
+                && newRules == this.Rules 
+                && newSuggestionModeItem == this.SuggestionModeItem)
+            {
+                return this;
+            }
+            else
+            {
+                return Create(newSpan, newItems, newRules, newSuggestionModeItem);
+            }
+        }
+
+        /// <summary>
+        /// Creates a copy of this <see cref="CompletionList"/> with the <see cref="DefaultSpan"/> property changed.
+        /// </summary>
+        public CompletionList WithDefaultSpan(TextSpan span)
```

Comment:
Naming.  You use "DefaultSpan" and "span" and "span:" these names should all be consistent.